src/util/codec/mysql/json/functions.rs

```diff
@@ -87,10 +108,157 @@ fn decode_escaped_unicode(s: &str) -> Result<char> {
     char::from_u32(u).ok_or(box_err!("invalid char from: {}", s))
 }
 
+// extract_json is used by JSON::extract().
+pub fn extract_json(j: &Json, path_legs: &[PathLeg]) -> Vec<Json> {
+    if path_legs.is_empty() {
+        return vec![j.clone()];
+    }
+    let (current_leg, sub_path_legs) = (&path_legs[0], &path_legs[1..]);
+    let mut ret = vec![];
+    match *current_leg {
+        PathLeg::Index(i) => {
+            match *j {
+                Json::Array(ref array) => {
+                    if i == PATH_EXPR_ARRAY_INDEX_ASTERISK {
+                        for child in array {
+                            ret.append(&mut extract_json(child, sub_path_legs))
+                        }
+                    } else if (i as usize) < array.len() {
+                        ret.append(&mut extract_json(&array[i as usize], sub_path_legs))
+                    }
+                }
+                _ => {
+                    if (i == PATH_EXPR_ARRAY_INDEX_ASTERISK) || (i as usize == 0) {
+                        ret.append(&mut extract_json(j, sub_path_legs))
+                    }
+                }
+            }
+        }
+        PathLeg::Key(ref key) => {
+            if let Json::Object(ref map) = *j {
+                if key == PATH_EXPR_ASTERISK {
+                    for key in map.keys() {
+                        ret.append(&mut extract_json(&map[key], sub_path_legs))
+                    }
+                } else if map.contains_key(key) {
+                    ret.append(&mut extract_json(&map[key], sub_path_legs))
+                }
+            }
+        }
+        PathLeg::DoubleAsterisk => {
+            ret.append(&mut extract_json(j, sub_path_legs));
+            match *j {
+                Json::Array(ref array) => {
+                    for child in array {
+                        ret.append(&mut extract_json(child, path_legs))
+                    }
+                }
+                Json::Object(ref map) => {
+                    for key in map.keys() {
+                        ret.append(&mut extract_json(&map[key], path_legs))
+                    }
+                }
+                _ => {}
+            }
+        }
+    }
+    ret
+}
+
 #[cfg(test)]
 mod test {
     use std::collections::BTreeMap;
     use super::*;
+    use super::super::path_expr::{PathExpressionFlag, PATH_EXPR_ARRAY_INDEX_ASTERISK,
+                                  PATH_EXPRESSION_CONTAINS_ASTERISK,
+                                  PATH_EXPRESSION_CONTAINS_DOUBLE_ASTERISK};
+
+    #[test]
+    fn test_json_extract() {
+        let mut m = BTreeMap::new();
+        m.insert(String::from("a"), Json::String(String::from("a1")));
```

Comment:
Since it's a must to quote the string like `"\"a1\""`, IMO it's simpler to the directly specify the Json type using `Json::String` constructor.